src/repr/src/adt/range.rs

```diff
@@ -0,0 +1,329 @@
+pub struct RangeBound<'a> {
+    pub inclusive: bool,
+    pub bound: Option<DatumNested<'a>>,
+}
+
+impl<'a> std::fmt::Display for RangeBound<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.bound {
+            None => Ok(()),
+            Some(bound) => write!(f, "{}", bound.datum()),
+        }
+    }
+}
+
+impl<'a> Ord for RangeBound<'a> {
+    /// `cmp` is only correct comparing lower bounds.
+    ///
+    /// - The output is correct for upper bounds if `Ordering::reverse`'d.
+    /// - Does not support comparing an upper and a lower bound.
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.bound.cmp(&other.bound) {
+            Ordering::Equal => {
+                if self.inclusive == other.inclusive {
+                    Ordering::Equal
+                } else if self.inclusive {
+                    Ordering::Less
+                } else {
+                    Ordering::Greater
+                }
+            }
+            o => o,
+        }
+    }
+}
+
+impl<'a> PartialOrd for RangeBound<'a> {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// A contrivance to make arguments to functions that construct ranges more
+/// well structured.
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, PartialOrd)]
+pub struct RangeBoundDesc {
+    // Is this bound infinite/unbounded?
+    pub inf: bool,
+    // Is this bound inclusive?
+    pub inc: bool,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum RangeError {
+    MisorderedRangeBounds,
+}
+
+impl fmt::Display for RangeError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            RangeError::MisorderedRangeBounds => {
+                f.write_str("range lower bound must be less than or equal to range upper bound")
+            }
+        }
+    }
+}
+
+impl Error for RangeError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        None
+    }
+}
+
+pub trait RangeOps<'a, T: Ord + PartialOrd + Eq + PartialEq>
+where
+    T: TryFrom<Datum<'a>>,
+    <T as TryFrom<Datum<'a>>>::Error: std::fmt::Debug,
+{
+    /// Reports whether this element is within the lower bound.
+    fn lower_elem_cmp(&self, elem: &T) -> bool;
+    /// Reports whether this element is within the upper bound.
+    fn upper_elem_cmp(&self, elem: &T) -> bool;
+    /// Reports whether this element is within the upper and lower bounds,
+    /// i.e. contained within range.
+    fn contains(&self, elem: &T) -> bool;
+}
+
+impl<'a, T: Ord + PartialOrd + Eq + PartialEq> RangeOps<'a, T> for Range<'a>
+where
+    T: TryFrom<Datum<'a>>,
+    <T as TryFrom<Datum<'a>>>::Error: std::fmt::Debug,
+{
+    fn lower_elem_cmp(&self, elem: &T) -> bool {
+        match self.inner {
+            None => false,
+            Some(inner) => match inner
+                .lower
+                .bound
+                .map(|bound| <T>::try_from(bound.datum()).unwrap())
```

Comment:
Can you replace this `unwrap` with an `unwrap_or_else` that prints the two values?